internal/controller/shortener.go

```diff
@@ -152,7 +261,241 @@ func (routes *shortenerRoutes) PostHandlerAPI(w http.ResponseWriter, r *http.Req
 		return
 	}
 
+	w.Header().Set("Content-Type", "application/json")
+	if isConflict {
+		w.WriteHeader(http.StatusConflict)
+	} else {
+		w.WriteHeader(http.StatusCreated)
+	}
+	w.Write(ShortURLResponse)
+}
+
+// PostBatch -.
+func (routes *shortenerRoutes) PostBatch(w http.ResponseWriter, r *http.Request) {
+	if r.Header.Get("Content-Type") != "application/json" {
+		http.Error(w, "only accepts Content-Type application/json", 400)
+		return
+	}
+
+	var reader io.Reader
+	if r.Header.Get(`Content-Encoding`) == `gzip` {
+		gz, err := gzip.NewReader(r.Body)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		reader = gz
+		defer gz.Close()
+	} else {
+		reader = r.Body
+	}
+
+	b, err := io.ReadAll(reader)
+	if err != nil {
+		http.Error(w, err.Error(), 400)
+		return
+	}
+
+	var IncomingMessage IncomingBatch
+	err = json.Unmarshal(b, &IncomingMessage)
+	if err != nil {
+		http.Error(w, err.Error(), 400)
+		return
+	}
+
+	if len(IncomingMessage) == 0 {
+		http.Error(w, "emty url list for shortening", 400)
+		return
+	}
+
+	var (
+		shortener     entity.Shortener
+		outgoingbatch OutGoingBatch
+	)
+
+	for _, v := range IncomingMessage {
+		userID, _, err := getUserID(r, w, v.LongURL, routes.Cfg)
+		if err != nil {
+			log.Printf("controller layer error: %v\n", err)
+		}
+
+		longURLString := v.LongURL
+		shortener, _, err = routes.sh.PostURLPair(context.Background(), entity.Shortener{
+			LongURL: longURLString,
+			UserID:  userID,
+		})
+		if err != nil {
+			log.Printf("controller layer error: %v\n", err)
+		}
+
+		var OutgoingPair OutgoingPairStruct
+		OutgoingPair.CorrelationID = v.CorrelationID
+		OutgoingPair.ShortUL = shortener.ShortURL
+		outgoingbatch = append(outgoingbatch, OutgoingPair)
+
+	}
+
+	ShortURLResponse, err := json.Marshal(outgoingbatch)
+	if err != nil {
+		http.Error(w, err.Error(), 400)
+		return
+	}
+
 	w.Header().Set("Content-Type", "application/json")
 	w.WriteHeader(http.StatusCreated)
 	w.Write(ShortURLResponse)
 }
+
+// getUserID -.
+func getUserID(r *http.Request, w http.ResponseWriter, URLString string, cfg *config.Config) (string, bool, error) {
```

Comment:
```suggestion
func getUserID(r *http.Request, w http.ResponseWriter, url string, cfg *config.Config) (string, bool, error) {
```
у тебя и так тип стринг, зачем это еще в названии указывать?

Reply:
переименовал